witness.md

```diff
@@ -77,617 +77,166 @@ It is also suitable to store subtries that can be used for semi-stateless
-
-- to the left of the `|=>` sign:
-
-    - optional `GUARD` statements;
+<Byte> := 0x00        {byte with value 0x00}
+        | 0x01        {byte with value 0x01}
+        | ...
+        | 0xff        {byte with value 0xff}
 
-    - the pattern to match against;
-
-- result, to the right of the `|=>` sign.
-
-
-#### `GUARD`s
-
-Each substitution rule can have zero, one or multiple `GUARD` statements.
-Each `GUARD` statement looks like this:
-
-```
-GUARD <CONDITION>
 ```
 
-For a substitution rule to be applicable, the `<CONDITION>` in its `GUARD` statement MUST be true.
-
-If a substitution rule has multiple `GUARD` statements, all of them MUST BE satisfied.
+First define some non-terminals to simplify later definitions.
 
-If there are no `GUARD` statements, the substitution rule's applicability is
-only defined by the PATTERN.
-
-Example:
-```
- GUARD NBITSET(mask) == 2
-|---- GUARD STATEMENT ---|
-
- Node(n0) Node(n1) BRANCH{mask} |=> 
- BranchNode{MAKE_VALUES_ARRAY(mask, n0, n1)}
 ```
+<U32> := u32:<Byte>^4		{u32 as a 32-bit unsigned integer in little-endian}
 
-For the example rule to be applicable both facts MUST be true:
-
-1. `mask` contains only 2 bits set to 1 (the rest are set to 0);
-
-2. to the left of `BRANCH` instruction there is at least 2 `Node`s.
-
-Fact (1) comes from the `GUARD` statement.
-
+<Bytes32> := b:<Byte>^32	{byte array b in big-endian}
 
-#### PATTERN
+<Address> := b:<Byte>^20	{byte array b in big-endian}
 
-`[NodeType(boundVar1)... NodeType(boundVarN)] Instruction[(param1... paramN)]`
+<Byte_Nonzero> := 0x01          {byte with value 0x01}
+                | 0x02          {byte with value 0x02}
+                | ...
+                | 0xff          {byte with value 0xff}
 
-A pattern MUST contain a single instruction.
-A pattern MAY contain one or more `Node`s to the left of the instruction to
-match.
-An instruction MAY have one or more parameters.
+<Byte_More_Than_One_Bit_Set> := 0x03          {byte with value 0x03}
+                              | 0x05          {byte with value 0x05}
+                              | 0x06          {byte with value 0x06}
+                              | 0x07          {byte with value 0x07}
+                              | 0x09          {byte with value 0x09}
+                              | 0x0a          {byte with value 0x0a}
+                              | ...
+                              | 0x0f          {byte with value 0x0f}
+                              | 0x11          {byte with value 0x11}
+                              | 0x12          {byte with value 0x12}
+                              | ...
+                              | 0xff          {byte with value 0xff}
 
-Pattern matching is happening by the types. `Node` type means any node is
-matched, some specific node type will require a specific match.
+<Bytes2_More_Than_One_Bit_Set> := b1:<Byte> b2case1:<Byte>^numbits(b1)>1 b2case2:<Byte_Nonzero>^numbits(b1)==1 b2case3:<Byte_More_Than_One_Bit_Set>^numbits(b1)==0
```

Comment:
I think this is missing some parentheses:
`b2case1:<Byte>^numbits(b1)>1` should be `b2case1:<Byte>^(numbits(b1)>1)`, right? At least that would be easier to read

Reply:
Yes, it may be easier to read like that. It was written this way because we already use all brackets `()[]{}<>`, so reusing parentheses without explaining notation may cause more confusion. It took me until #35 to add rough notation for expressions, which finally allows parentheses for things like this.

This problem could be indirectly solved if this document were rendered to latex. Then it would be `b2case1:<Byte>^{numbits(b1)>1}`.

This problem could also be indirectly solved by putting complexity into the validation rules -- having a rule for each case, and choosing the rule based on the validation section. But that moves away from a (naive) goal of expressing as much as possible in syntax.

I will put this on my todo list. But I am now focused on stabilizing the binary encoding. Aesthetics can will be addressed later.